src/interchange/src/avro.rs

```diff
@@ -1987,23 +1984,42 @@ impl SchemaCache {
-    async fn get(&mut self, id: i32, reader_schema: &Schema) -> anyhow::Result<Option<&Schema>> {
+    /// that this schema cache was initialized with, returns the schema directly.
+    /// If not, performs schema resolution on the reader and writer and
+    /// returns the result.
+    async fn get(&mut self, id: i32, reader_schema: &Schema) -> anyhow::Result<&Schema> {
         match self.cache.entry(id) {
-            Entry::Occupied(o) => Ok(o.into_mut().as_ref()),
+            Entry::Occupied(o) => o
+                .into_mut()
+                .as_ref()
+                .map_err(|e| anyhow::Error::new(e.clone())),
             Entry::Vacant(v) => {
-                // TODO(benesch): make this asynchronous, to avoid blocking the
-                // Timely thread on this network request.
-                let res = self.ccsr_client.get_schema_by_id(id).await?;
-                let schema = parse_schema(&res.raw)?;
-                if schema.fingerprint::<Sha256>().bytes == self.reader_fingerprint.bytes {
-                    Ok(v.insert(None).as_ref())
-                } else {
-                    // the writer schema differs from the reader schema,
-                    // so we need to perform schema resolution.
-                    let resolved = resolve_schemas(&schema, reader_schema)?;
-                    Ok(v.insert(Some(resolved)).as_ref())
-                }
+                // An issue with _fetching_ the schema should be returned
+                // immediately, and not cached, since it might get better on the
+                // next retry.
+                // TODO - some sort of exponential backoff or similar logic
+                let response = self.ccsr_client.get_schema_by_id(id).await?;
+                // Now, we've gotten some json back, so we want to cache it (regardless of whether it's a valid
+                // avro schema, it won't change).
+                //
+                // However, we can't just cache it directly, since resolving schemas takes significant CPU work,
+                // which  we don't want to repeat for every record. So, parse and resolve it, and cache the
+                // result (whether schema or error).
+                let rf = &self.reader_fingerprint.bytes;
+                let result = (|| {
+                    let schema = Schema::parse_str(&response.raw)?;
+                    if &schema.fingerprint::<Sha256>().bytes == rf {
+                        Ok(schema)
+                    } else {
+                        // the writer schema differs from the reader schema,
+                        // so we need to perform schema resolution.
+                        let resolved = resolve_schemas(&schema, reader_schema)?;
+                        Ok(resolved)
+                    }
+                })();
+                v.insert(result)
+                    .as_ref()
+                    .map_err(|e| anyhow::Error::new(e.clone()))
             }
```

Comment:
I think this small reorg is slightly simpler, if you like:

```rust
    async fn get(&mut self, id: i32, reader_schema: &Schema) -> anyhow::Result<&Schema> {
        let entry = match self.cache.entry(id) {
            Entry::Occupied(o) => o.into_mut(),
            Entry::Vacant(v) => {
                // An issue with _fetching_ the schema should be returned
                // immediately, and not cached, since it might get better on the
                // next retry.
                // TODO - some sort of exponential backoff or similar logic
                let response = self.ccsr_client.get_schema_by_id(id).await?;
                // Now, we've gotten some json back, so we want to cache it (regardless of whether it's a valid
                // avro schema, it won't change).
                //
                // However, we can't just cache it directly, since resolving schemas takes significant CPU work,
                // which  we don't want to repeat for every record. So, parse and resolve it, and cache the
                // result (whether schema or error).
                let rf = &self.reader_fingerprint.bytes;
                let result = Schema::parse_str(&response.raw).and_then(|schema| {
                    if &schema.fingerprint::<Sha256>().bytes == rf {
                        Ok(schema)
                    } else {
                        // the writer schema differs from the reader schema,
                        // so we need to perform schema resolution.
                        resolve_schemas(&schema, reader_schema)
                    }
                });
                v.insert(result)
            }
        };
        entry.as_ref().map_err(|e| anyhow::Error::new(e.clone()))
    }
```